File: tools/csp_server.js
const http = require('http');
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

const ROOT = path.resolve(__dirname, '..');
const PORT = 8001;

function send404(res) {
  res.writeHead(404);
  res.end('Not found');
}

function sendFile(res, filePath, cspHeader = null) {
  fs.readFile(filePath, (err, data) => {
    if (err) return send404(res);
    const ext = path.extname(filePath).toLowerCase();
    const ct = ext === '.html' ? 'text/html; charset=utf-8' : (ext === '.js' ? 'application/javascript' : (ext === '.css' ? 'text/css' : 'application/octet-stream'));
    const headers = { 'Content-Type': ct };
    if (cspHeader) headers['Content-Security-Policy'] = cspHeader;
    res.writeHead(200, headers);
    res.end(data);
  });
}

const server = http.createServer((req, res) => {
  const url = new URL(req.url, `http://${req.headers.host}`);
  let pathname = decodeURIComponent(url.pathname);
  if (pathname === '/' || pathname === '/index.html') {
    const indexPath = path.join(ROOT, 'index.html');
    fs.readFile(indexPath, 'utf8', (err, data) => {
      if (err) return send404(res);
      const nonce = crypto.randomBytes(16).toString('base64');
      const csp = `script-src 'self' 'nonce-${nonce}' https://static.cloudflareinsights.com; object-src 'none'; base-uri 'self';`;
      const body = data.replace(/%CSP_NONCE%/g, nonce);
      res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8', 'Content-Security-Policy': csp });
      res.end(body);
    });
    return;
  }

  const filePath = path.join(ROOT, pathname);
  fs.stat(filePath, (err, stats) => {
    if (err || !stats.isFile()) return send404(res);
    sendFile(res, filePath);
  });
});

server.listen(PORT, () => console.log(`CSP server listening on http://localhost:${PORT}`));

process.on('SIGINT', () => {
  console.log('\nShutting down');
  server.close();
  process.exit(0);
});

File: tools/csp_server.py
#!/usr/bin/env python3
"""Simple static file server that injects a per-request CSP nonce into index.html
and sets a header-based CSP allowing scripts from 'self' and the nonce.
"""
import http.server
import socketserver
import base64
import os
import logging
import traceback
from urllib.parse import unquote, urlparse

PORT = 8001
ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))

# Enable basic logging
logging.basicConfig(level=logging.INFO, format='[CSP SERVER] %(message)s')

class CSPRequestHandler(http.server.SimpleHTTPRequestHandler):
    def translate_path(self, path):
        # Serve files relative to project root
        path = urlparse(path).path
        path = unquote(path)
        if path == '/' or path == '/index.html':
            return os.path.join(ROOT, 'index.html')
        return os.path.join(ROOT, path.lstrip('/'))

    def end_headers(self):
        # If serving index, generate nonce and set CSP header
        if self.path == '/' or self.path == '/index.html':
            nonce = base64.b64encode(os.urandom(16)).decode('ascii')
            # Separate rules: inline scripts via nonce; external script elements allowed from Cloudflare
            csp = (
                "script-src 'self' 'nonce-" + nonce + "'; "
                "script-src-elem 'self' https://static.cloudflareinsights.com; "
                "object-src 'none'; base-uri 'self';"
            )
            self.send_header('Content-Security-Policy', csp)
            # Also expose nonce to downstream (not required, we inject into body)
            self.server.current_nonce = nonce
            logging.info(f"Set CSP header on {self.path}: {csp}")
        super().end_headers()

    def send_head(self):
        path = self.translate_path(self.path)
        if os.path.isdir(path):
            for index in ("index.html",):
                index_path = os.path.join(path, index)
                if os.path.exists(index_path):
                    path = index_path
                    break
        if path.endswith('index.html') and os.path.exists(path):
            # Read and inject nonce placeholder
            try:
                with open(path, 'rb') as f:
                    data = f.read().decode('utf-8')
            except Exception:
                return http.server.SimpleHTTPRequestHandler.send_head(self)
            # Ensure nonce is generated (end_headers sets it)
            nonce = base64.b64encode(os.urandom(16)).decode('ascii')
            # Include script-src-elem to explicitly allow external <script> elements
            csp = (
                "script-src 'self' 'nonce-" + nonce + "' https://static.cloudflareinsights.com; "
                "script-src-elem 'self' 'nonce-" + nonce + "' https://static.cloudflareinsights.com; "
                "object-src 'none'; base-uri 'self';"
            )
            self.send_response(200)
            self.send_header('Content-type', 'text/html; charset=utf-8')
            self.send_header('Content-Security-Policy', csp)
            logging.info(f"Injected nonce into index and set CSP: {csp}")
            self.end_headers()
            body = data.replace('%CSP_NONCE%', nonce)
            return io.BytesIO(body.encode('utf-8'))
        return http.server.SimpleHTTPRequestHandler.send_head(self)

if __name__ == '__main__':
    import io
    Handler = CSPRequestHandler
    # Allow quick restarts on the same port
    socketserver.TCPServer.allow_reuse_address = True
    with socketserver.TCPServer(("", PORT), Handler) as httpd:
        logging.info(f"Serving on port {PORT} with header-based CSP and per-request nonce")
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            logging.info('Shutting down (KeyboardInterrupt)')
            httpd.server_close()
        except Exception as e:
            logging.info('Server crashed with exception:')
            traceback.print_exc()
            httpd.server_close()
